components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -0,0 +1,291 @@
+//
+// The parsed uint256 value is serialised as a hex string prefixed by "0x".
+bool GetUint256HexFromData(const std::string& input,
+                           std::string* arg,
+                           size_t offset) {
+  CHECK(arg);
+
+  std::string padded_arg = "0x" + input.substr(offset, 64);
+  uint256_t arg_uint;
+  if (!brave_wallet::HexValueToUint256(padded_arg, &arg_uint))
+    return false;
+
+  *arg = brave_wallet::Uint256ValueToHex(arg_uint);
+  return true;
+}
+
+// GetBoolFromData extracts a 32-byte wide boolean value from the
+// calldata at the specified offset.
+//
+// The parsed bool value is serialised as "true" or "false" strings.
+bool GetBoolFromData(const std::string& input,
+                     std::string* arg,
+                     size_t offset) {
+  CHECK(arg);
+
+  std::string padded_arg = "0x" + input.substr(offset, 64);
+  uint256_t arg_uint;
+  if (!brave_wallet::HexValueToUint256(padded_arg, &arg_uint))
+    return false;
+
+  if (arg_uint == 0)
+    *arg = "false";
+  else if (arg_uint == 1)
+    *arg = "true";
+  else
+    return false;
+
+  return true;
+}
+
+// GetBytesHexFromData extracts a bytes value from the calldata at the
+// specified offset using head-tail encoding mechanism. bytes are packed
+// tightly in chunks of 32 bytes, with the first 32 bytes encoding the length,
+// followed by the actual content.
+//
+// The parsed bytearray is serialised as a hex string prefixed by "0x".
+bool GetBytesHexFromData(const std::string& input,
+                         std::string* arg,
+                         size_t offset) {
+  CHECK(arg);
+
+  size_t pointer;
+  if (!GetSizeFromData(input, &pointer, offset))
+    return false;
+
+  size_t bytes_len;
+  if (!GetSizeFromData(input, &bytes_len, pointer * 2))
+    return false;
+
+  *arg = "0x" + input.substr(pointer * 2 + 64, bytes_len * 2);
+  return true;
+}
+
+// GetAddressArrayFromData parses a calldata sequence to extract a dynamic
+// array of addresses at the specified offset using head-tail encoding
+// mechanism. The encoding is similar to bytes, with the first 32 bytes
+// representing the number of elements in the array, followed by each element.
+//
+// The parsed data is joined together into a hex string prefixed by "0x".
+bool GetAddressArrayFromData(const std::string& input,
+                             std::string* arg,
+                             size_t offset) {
+  CHECK(arg);
+
+  size_t pointer;
+  if (!GetSizeFromData(input, &pointer, offset))
+    return false;
+
+  size_t array_len;
+  if (!GetSizeFromData(input, &array_len, pointer * 2))
+    return false;
+
+  size_t array_offset = pointer * 2 + 64;
+  *arg = "0x";
+  for (size_t i = 0; i < array_len; i++) {
+    std::string value;
+    if (!GetAddressFromData(input, &value, array_offset))
+      return false;
+    *arg += value.substr(2);
+    array_offset += 64;
+  }
+
+  return true;
+}
+
+bool ABIDecodeInternal(const std::vector<std::string> types,
```

Comment:
& is missing

Reply:
Fixed 4b8c0e7eab2ea2babca5e9f1c3e9fb4834292f66.